### ISSUE TITLE
Allow assigning a name to a job

### ISSUE THREAD

**Issue by nonhermitian** [Qiskit/qiskit-ibmq-provider#118]
### What is the expected behavior?
Moved from: https://github.com/Qiskit/qiskit-terra/issues/2406

As I have mentioned before (but cannot find the Issue for), we really need the ability to give a job a name (`job.name`), and to be able to query the database by this name, or a subset of characters in this name. The job id is good as a unique identifier, but less than ideal for actually labeling results for future retrieval. It would also be nice to set the name of a job after it is run, so that if I have good results I can say: job.set_name('These are good results').

**Comment by nonhermitian** [comment]
To quote @jaygambetta's last comment:

> I agree we should do this.

> job.name --> returns the name
> job.set_name() = "blah" (do we need to have a set?)

> one question I have that we have punted on but I think we need to address before we do this is do > the job options gets associated with a backend or a provider.

> currently

> backend.jobs() - returns a list of previous jobs executed by this user on this backend.
> backend.retrieve_job() - returns a job by a job_id.

> but I think this actually does not do it by the backend (@diego-plan9) is this correct? It really does it > by the provider. If so should this be

> provider.jobs() - returns a list of previous jobs executed by this user on this backend.
> provider.retrieve_job() - returns a job by a job_id.

> and then the

> job = provider.search(name = "name")

**Comment by nonhermitian** [comment]
To answer @jaygambetta's question, the backend itself makes the query to the API for its jobs, not the provider.

**Comment by diego-plan9** [comment]
This would likely need to coordinate with the API team, and probably needs to be tackled after #104 to set all the pieces in place. One of the underlying topics is that the concept of a `Job` in Qiskit is not too heavily mapped to what a `Job` is for the API - ideally, we should make an effort for reconciling them and formalize them to a degree (ie. specs!) _before_ adding extra fields and functions, in order to try to avoid further disparities and their related side effects.

**Comment by pacomf** [comment]
@nonhermitian the name needs to be unique in all jobs for an user? Or can be the same and you can return several jobs with the same name?

**Comment by nonhermitian** [comment]
The job ID is the unique identifier.  The job name need not be.

**Comment by pacomf** [comment]
ok!

**Comment by nonhermitian** [comment]
@pacomf said that the API will already allow this, so I think the API action label can be removed.

**Comment by ajavadia** [comment]
I agree this is extremely nice to have. Currently it is hard to reverse engineer what a job was supposed to do.

**Comment by diego-plan9** [comment]
After talking with the API team and taking into account the current status of the provider, I think we can add an initial implementation, albeit limited in scope. In particular, allowing:
* setting a custom job name when submitting a job via `backend.run()`
* filtering by that custom job name when getting jobs back via `backend.jobs()`

Hopefully this will allow for a minimal yet functional feature (in the sense of allowing users to start setting the name for their submitted jobs and retrieve them back), while working on #104 to allow the rest of the features to be added, such as the ideas on  https://github.com/Qiskit/qiskit-ibmq-provider/issues/118#issuecomment-495161766 (which some of them might require help from the API).

As hinted in https://github.com/Qiskit/qiskit-ibmq-provider/issues/118#issuecomment-497319130, the `Job` class is in need of an overhaul (and is arguably the next "big feature") that will likely imply some tweaks to its public API - and in general I think we are better off holding off changes to it until #104 gets completed in order to both avoid the risk of introducing public API modifications that might be short-lived, and also due to the inconveniences when working with the current class.

**Comment by nonhermitian** [comment]
This works for me.

**Comment by diego-plan9** [comment]
Closing after #384.
